### PR TITLE
Documenting CA certificates setup for hetty

### PR DESCRIPTION
**Pull request by tcarrio**
- Provides more accessible instructions on CA certificates
- New sections provided for "Certificate Setup and Installation"
- Configuring certificates for hetty using defaults and direct CLI
  arguments
- Trusting CA certificates on Windows, macOS, and Ubuntu
- Mention for additional Linux distros not covered by Ubuntu

Ref: #12

### CODE REVIEW

**Comment by tcarrio** [comment]
I just went through this process myself getting a local dev environment up and saw the note in the README, so I figured I would include some documentation.

This was only tested on Linux, but I included instructions for Ubuntu, macOS, and Windows by referencing various guides. I hope they're helpful 👍

File: README.md [hunk 1]
```diff
@@ -86,8 +86,67 @@ Usage of ./hetty:
...
+_command line utilities. This process was done on a Linux machine but should_
+_provide guidance on Windows and macOS as well._
+
+### Generating CA certificates
+
+You can start off by generating a new key and CA certificate which will both expire
+after a month.
+
+```sh
+openssl req -newkey rsa:2048 -new -nodes -x509 -days 31 -keyout hetty_key.pem -out hetty_cert.pem
```
Comment (RiRa12621) [inline README.md:107]: I don't think that's needed anymore.
Hetty generates a set of certs on startup

Comment (tcarrio) [reply README.md:107]: Good point! I can make a mention of the automatically generated certificates as well 👍

Comment (tcarrio) [reply README.md:127]: Latest commit 9a8afda now mentions instructions for both Hetty and OpenSSL, mentioning the benefit of one over the other, and suggesting Hetty defaults for beginners. More advanced users who need control over certificate expiration or cryptography strength can use OpenSSL if they would like, and the instructions outline the process.

File: README.md [hunk 2]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
         CA private key filepath. Creates a new CA private key if file doesn't exist (default "~/.hetty/hetty_key.pem")
 ```
 
-⚠️ _Todo: Write instructions for installing CA certificate in local CA store, and_
-_configuring Hetty to be used as a proxy server._
+## Certificate Setup and Installation
+
+In order for Hetty to proxy requests going to HTTPS endpoints, certificates for
```
Comment (dstotijn) [inline README.md:91]:
```suggestion
In order for Hetty to proxy requests going to HTTPS endpoints, a root CA certificate for 
```

File: README.md [hunk 3]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
         CA private key filepath. Creates a new CA private key if file doesn't exist (default "~/.hetty/hetty_key.pem")
 ```
 
-⚠️ _Todo: Write instructions for installing CA certificate in local CA store, and_
-_configuring Hetty to be used as a proxy server._
+## Certificate Setup and Installation
+
+In order for Hetty to proxy requests going to HTTPS endpoints, certificates for 
+Hetty will need to be set up. Furthermore, these certificates may need to be
```
Comment (dstotijn) [inline README.md:92]:
```suggestion
Hetty will need to be set up. Furthermore, the CA certificate may need to be
```

File: README.md [hunk 4]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
...
+will cover how you can generate your certificate, provide them to hetty, and how
+you can install them in your local CA store.
+
+⚠️ _This process was done on a Linux machine but should_
+_provide guidance on Windows and macOS as well._
+
+### Generating CA certificates
+
+You can generate CA certificates two different ways. The first is bundled directly
+with hetty, and simplifies the process immensely. The alternative is using OpenSSL
```
Comment (dstotijn) [inline README.md:103]:
```suggestion
with Hetty, and simplifies the process immensely. The alternative is using OpenSSL
```

File: README.md [hunk 5]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
...
+installed to the host for them to be trusted by your browser. The following steps
+will cover how you can generate your certificate, provide them to hetty, and how
+you can install them in your local CA store.
+
+⚠️ _This process was done on a Linux machine but should_
+_provide guidance on Windows and macOS as well._
+
+### Generating CA certificates
+
+You can generate CA certificates two different ways. The first is bundled directly
```
Comment (dstotijn) [inline README.md:102]:
```suggestion
You can generate a CA keypair two different ways. The first is bundled directly
```

File: README.md [hunk 6]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
...
+You should now have a key and certificate located at  `~/.hetty/hetty_key.pem` and
+`~/.hetty/hetty_cert.pem` respectively.
+
+#### Generating CA certificates with OpenSSL
+
+You can start off by generating a new key and CA certificate which will both expire
+after a month.
+
+```sh
+openssl req -newkey rsa:2048 -new -nodes -x509 -days 31 -keyout hetty_key.pem -out hetty_cert.pem
```
Comment (dstotijn) [inline README.md:127]: Maybe simplify this section by creating the keypair directly in `~/.hetty`.
```suggestion
mkdir ~/.hetty
openssl req -newkey rsa:2048 -new -nodes -x509 -days 31 -keyout ~/.hetty/hetty_key.pem -out ~/.hetty/hetty_cert.pem
```

File: README.md [hunk 7]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
...
+
+```
+hetty -key key.pem -cert cert.pem
+```
+
+### Installing CA certificates
+
+In order for your browser to allow traffic to the local Hetty proxy, you may need
+to install these certificates to your local CA store. Otherwise, you may be met
+with errors related to HTTP Strict Transport Security (HSTS). In the real world this
```
Comment (dstotijn) [inline README.md:143]: I don't this HSTS is relevant here in this context; this is about trust for the CA we've generated. We could maybe rephrase this paragraph to:

> Your browser first needs to trust the CA certificate we generated earlier. By default it doesn't, and any certificates signed by Hetty during proxying using the CA will be treated as invalid.

File: README.md [hunk 8]
```diff
@@ -86,8 +86,87 @@ Usage of ./hetty:
...
+The default location which `hetty` will check for the key and CA certificate is under
+`~/.hetty/`, at `hetty_key.pem` and `hetty_cert.pem` respectively. You can move them
+here and `hetty` will detect them automatically. Otherwise, you can specify the
+location of these as arguments to `hetty`.
+
+```
+hetty -key key.pem -cert cert.pem
+```
+
+### Installing CA certificates
```
Comment (dstotijn) [inline README.md:139]:
```suggestion
### Trusting the CA certificate
```

**Review by dstotijn — CHANGES_REQUESTED** [review]
Thanks for taking the effort @tcarrio! Left a few comments, but in general I think this is a good start. We'll probably refactor the certificate instructions into a separate docs site, but for now I think the README is the best place.

**Comment by elithrar** [comment]
You should consider building on, or using, mkcert: https://github.com/FiloSottile/mkcert

It automates the installation of the CA on major OS’ and is well-tested.

**Comment by tcarrio** [comment]
All of the comments look valid. I'll be applying all of them, which I can't seem to batch as a single commit in the GitHub UI. I will apply the updates and push again next chance.

File: README.md (hunk 7 above)
Comment (tcarrio) [reply README.md:143]: The reason I included this was because I had tested this case myself in Firefox and ran into this error, and thought it may be useful for others who are confused why it's not working. I can removed this part if you would like.

Comment (tcarrio) [reply README.md]: This has been removed as it may be misleading or irrelevant.

**Comment by tcarrio** [comment]
This has been updated according to review. Please let me know what else may need updating 👍

**Comment by Lawri-van-Buel** [comment]
> You should consider building on, or using, mkcert: https://github.com/FiloSottile/mkcert
> 
> It automates the installation of the CA on major OS’ and is well-tested.
I successfully used the rot certs from mkcert as the source CA file for Hetty.

I also completely agree that we should  employ the concept of doing 1 thing, but doing it really good. makign / managing a CA should be out of scope for Hetty... but it could include instructions / inclusion of mkcert for that purpose.

As to the instructions using mkcert: [github.com/FiloSottile/mkcert](https://github.com/FiloSottile/mkcert)
> 1. First either install the golang build environmnet from [golang.org](https://golang.org/) or skip to step 4
> 1. Get a local checkout of the source from https://github.com/FiloSottile/mkcert.git make sure you have a taged version of the code by checking out a taged release (get the list of tags with `git tag -l` and  than use the latest tag with `git checkout <tagname>`. if you are not changing the code tyourself you can ignore the detached message, otherwise add a branch with the `-b <branch name>` added to the last command.
> 1. Build mkcert using the following set of instructions. `go mod vendor` and `go build -ldflags "-X main.Version=$(git describe --tags)"`. This will yield the executable of mkcert for your system.
> 1. the mkcert will need to generate its CA files before it can generate any other certs, if you execute it likw `mkcert -install` will both generate the CA certificate & key and install it into all relevant (local) certificate stores (the one form your OS, the one from firefox, etc).
> 1. after that you can get the location of the Certificates file with `mkcert -CAROOT` it will return the directory where mkcert installed the certificates, you can *COPY* those into the hetty config as the CAcert and CAkey files.


I hope that helps, and if you like I am willign to also add it to the readme in a PR (after this one is accepted).

File: README.md [hunk 9]
```diff
@@ -86,8 +86,85 @@ Usage of ./hetty:
...
-_configuring Hetty to be used as a proxy server._
+## Certificate Setup and Installation
+
+In order for Hetty to proxy requests going to HTTPS endpoints, a root CA certificate for 
+Hetty will need to be set up. Furthermore, the CA certificate may need to be
+installed to the host for them to be trusted by your browser. The following steps
+will cover how you can generate your certificate, provide them to hetty, and how
+you can install them in your local CA store.
+
+⚠️ _This process was done on a Linux machine but should_
```
Comment (RiRa12621) [inline README.md:97]: I wonder if this should be native markdown `:warning:` instead of an embedded emoji

**Review by RiRa12621 — APPROVED** [review]
LGTM since i'm not super opinionated about the emoji i mentioned.
Squash and then it's good to go IMHO.
However I'd hold of with squashing until you got LGTM from @dstotijn

**Comment by dstotijn** [comment]
> I also completely agree that we should employ the concept of doing 1 thing, but doing it really good. makign / managing a CA should be out of scope for Hetty... but it could include instructions / inclusion of mkcert for that purpose.
> 
> (...)
>
> I hope that helps, and if you like I am willign to also add it to the readme in a PR (after this one is accepted).

@Lawri-van-Buel: I think the way forward is to use [mkcert](https://github.com/FiloSottile/mkcert) as a _library_ (import/dependency) in Hetty, at least for installing a CA in a system/browser CA store. That way we can leverage mkcert's features for CA management without introducing a lot (arguably non-core) code in Hetty. The alternative of instructing users to manually download/use `mkcert` to generate a keypair and install it is a bit of a degraded user experience; especially the generating of the keypair should remain a feature of Hetty itself, just like `mitmproxy`, Burp Suite and OWASP ZAP have this.

For now, generating the CA keypair ourselves and instructing users to install to their system/browser CA store is fine, and leading up to an eventual v1.0 release we can work on embedding mkcert. I'll create a separate issue for this in due time, but if you want to work on it already or discuss/how to do it, that'd be much appreciated! Probably [discussions](https://github.com/dstotijn/hetty/discussions) would be the best place.

**Review by dstotijn — APPROVED** [review]
LGTM! 👍
